src/Http/Routing/src/RouteGroupBuilder.cs

```diff
@@ -78,12 +86,25 @@ public void Dispose()
             }
         }
 
-        private RouteGroupContext GetNextRouteGroupContext(RoutePattern? prefix, IReadOnlyList<Action<EndpointBuilder>> conventions, IServiceProvider applicationServices)
+        private RouteGroupContext GetNextRouteGroupContext(
+            RoutePattern? prefix,
+            IReadOnlyList<Action<EndpointBuilder>> conventions,
+            IReadOnlyList<Action<EndpointBuilder>> finallyConventions,
+            IServiceProvider applicationServices)
         {
             var fullPrefix = RoutePatternFactory.Combine(prefix, _routeGroupBuilder._partialPrefix);
             // Apply conventions passed in from the outer group first so their metadata is added earlier in the list at a lower precedent.
             var combinedConventions = RoutePatternFactory.CombineLists(conventions, _routeGroupBuilder._conventions);
-            return new RouteGroupContext(fullPrefix, combinedConventions, applicationServices);
+            finallyConventions.Reverse();
+            _routeGroupBuilder._finallyConventions.Reverse();
```

Comment:
Just calling `GetGroupedEndpoints()` shouldn't mutate the state. `GetGroupedEndpoints()` could be called multiple times. We should add a test where we add multiple levels of group finally conventions and call `GetGroupedEndpoints()` multiple times and make sure they all run in the right order.

Reply:
Good call. Added a test to validate idempotency of this operation.